stdlib/toolchain/Compatibility50/BuiltinProtocolConformances.cpp

```diff
@@ -0,0 +1,92 @@
+//===--- BuiltinProtocolWitnessTables.cpp ---------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Definitions of some builtin protocol witness tables for backward compat.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Overrides.h"
+#include "swift/Runtime/BuiltinProtocolConformances.h"
+#include "swift/Runtime/Casting.h"
+#include "swift/Runtime/Debug.h"
+
+using namespace swift;
+
+extern const ProtocolDescriptor
+PROTOCOL_DESCRIPTOR_SYM(SWIFT_EQUATABLE_MANGLING);
+
+extern const ProtocolConformanceDescriptor
+BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_SYM(VARIADIC_TUPLE_MANGLING,
+                                            SWIFT_EQUATABLE_MANGLING);
+
+SWIFT_RUNTIME_EXPORT
+const _WitnessTable<1> swift::
+BUILTIN_PROTOCOL_WITNESS_TABLE_SYM(VARIADIC_TUPLE_MANGLING,
+                                   SWIFT_EQUATABLE_MANGLING) = {
+  &BUILTIN_PROTOCOL_CONFORMANCE_DESCRIPTOR_SYM(VARIADIC_TUPLE_MANGLING,
+                                               SWIFT_EQUATABLE_MANGLING),
+  {
+    reinterpret_cast<void *>(
+      BUILTIN_PROTOCOL_WITNESS_SYM(VARIADIC_TUPLE_MANGLING,
+                                   SWIFT_EQUATABLE_MANGLING,
+                                   SWIFT_EQUAL_OPERATOR_MANGLING))
+  }
+};
+
+SWIFT_RUNTIME_EXPORT SWIFT_CC(swift)
+bool swift::
+BUILTIN_PROTOCOL_WITNESS_SYM(VARIADIC_TUPLE_MANGLING,
+                             SWIFT_EQUATABLE_MANGLING,
+                             SWIFT_EQUAL_OPERATOR_MANGLING)
+(OpaqueValue *tuple1, OpaqueValue *tuple2, SWIFT_CONTEXT Metadata *swiftSelf,
+ Metadata *Self, void *witnessTable) {
+  auto tuple = cast<TupleTypeMetadata>(Self);
+
+  // Loop through all elements, and check if both tuples element is equal.
+  for (size_t i = 0; i != tuple->NumElements; i += 1) {
+    auto elt = tuple->getElement(i);
+
+    // Ensure we actually have a conform to Equatable for this element type.
+    auto equatable = &PROTOCOL_DESCRIPTOR_SYM(SWIFT_EQUATABLE_MANGLING);
+    auto conformance = swift50override_conformsToProtocol(elt.Type, equatable,
+                                                      swift_conformsToProtocol);
+
+    // If we don't have a conformance then something somewhere messed up in
+    // deciding that this tuple type was Equatable...??
+    if (!conformance)
+      abort();
```

Comment:
`fatalError` seems to be missing from these compatibility libraries?

Reply:
The compatibility libraries might in some cases be linked into a binary that doesn't itself use `libswiftCore`, such as a main app binary that is all C/ObjC/C++ but which transitively uses some library containing Swift that needs the compatibility binary, so it can't directly use Swift runtime facilities. For something like `fatalError`, adding a local version that logs a message and calls `abort` should be fine. (More generally, you might need to "soft link" Swift runtime functionality by using dlopen/dlsym.)